git_machete/git_operations.py

```diff
@@ -812,10 +819,48 @@ def is_equivalent_tree_reachable(
         )
 
         result = earlier_tree_hash in intermediate_tree_hashes
-        debug(f"result = {result}")
-        self.__is_equivalent_tree_reachable_cached[equivalent_to_commit_hash, reachable_from_commit_hash] = result
+        debug(f"intermediate tree hashes result = {result}")
+
+        if result or opt_squash_merge_detection != SquashMergeDetection.EXACT:
+            cache_value = (opt_squash_merge_detection, result)
+            self.__is_equivalent_tree_reachable_cached[equivalent_to_commit_hash, reachable_from_commit_hash] = cache_value
+            return result
+
+        # Let's try another way, a little more complex but takes into account the possibility
+        # that there were other commits between the common ancestor of the two branches and the squashed merge.
+        common_ancestor = self.get_merge_base(reachable_from_commit_hash, equivalent_to_commit_hash)
+        if not common_ancestor:
+            return False
+
+        equivalent_changeset = self._popen_git(
+            "diff",
+            common_ancestor,
+            equivalent_to_commit_hash
+        ).stdout
+        if equivalent_changeset.strip() == '':
+            # Empty changeset means the branches are identical, so the tree is equivalent.
+            cache_value = (opt_squash_merge_detection, True)
+            self.__is_equivalent_tree_reachable_cached[equivalent_to_commit_hash, reachable_from_commit_hash] = cache_value
+            return cache_value[1]
+
+        equivalent_patch_id = self.get_patch_id(equivalent_changeset)
+        patchmap = self.get_patch_ids_for_commits_between(
+            common_ancestor, reachable_from_commit_hash, MAX_COMMITS_FOR_SQUASH_MERGE_DETECTION)
+        patch_ids = set(patchmap.values())
+        result = equivalent_patch_id in patch_ids
+
+        debug(f"patch id result = {result}")
+        cache_value = (opt_squash_merge_detection, result)
+        self.__is_equivalent_tree_reachable_cached[equivalent_to_commit_hash, reachable_from_commit_hash] = cache_value
         return result
 
+    def get_patch_id(self, patch_contents: str) -> Optional[str]:
+        out = utils.get_non_empty_lines(self._popen_git("patch-id", input=patch_contents).stdout)
+
+        if len(out) == 0:
+            return None
```

Comment:
Line uncovered by tests. Probably good to leave it so, there doesn't seem to be any path that would cover this case. Still, doesn't feel right to just `# pragma: no cover` it away.